### ISSUE TITLE
Clippy not detecting all problems in multi-crate project (v2)

### ISSUE THREAD

**Issue by bobhy** [rust-lang/rust-clippy#11414]
### Summary

(Originally opened as https://github.com/rust-lang/cargo/issues/12561 .  Cargo team thinks this is a clippy issue.)

### Problem

I have a multi-crate project and want lint `clippy::wildcard_dependencies` enabled for all crates.
To configure this, I add `build.rustflags = ["-Dclippy::wildcard_dependencies"] to the top-level `./.cargo/config.toml` (there being no hierarchical config for clippy, currently).

This config does complain as desired about wildcard dependencies in child crates, but fails to complain about wildcard dependencies in the parent. 

If I change the project's root `Cargo.toml` so the child crates are not in the workspace, `cargo clippy` detects wildcard dependencies in the root crate (but not, of course in the children). 

And finally, if I eliminate wildcard dependencies from all the the child crates, `cargo clippy` fails to complain about the remaining wildcard dependency in the parent crate.  This is the crux of the bug report.

I'm thinking this is more a cargo problem than a clippy problem, or, of course, that `build.rustflags` doesn't do what I want it to do.

### Reproducer

1. `git clone https://github.com/bobhy/wc_dep; cd wc_dep`
2. `cp Cargo_fails.toml Cargo.toml`
3. `cp crates/sub_pop/Cargo_wc_dep.toml crates/sub_pop/Cargo.toml` # child crate *has* a wildcard dependency
4. 
```
> cargo clippy
    Checking sub_pop v0.1.0 (/home/bobhy/src/rust/wc_dep/crates/sub_pop)
error: wildcard dependency for `test_crate_test`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#wildcard_dependencies
  = note: requested on the command line with `-D clippy::wildcard-dependencies`

error: could not compile `sub_pop` (lib) due to previous error
```
Note: there is an additional wildcard dependency in the root `./Cargo.toml` that was not detected!

### Version

```text
rustc 1.71.1 (eb26296b5 2023-08-03)
binary: rustc
commit-hash: eb26296b556cef10fb713a38f3d16b9886080f26
commit-date: 2023-08-03
host: x86_64-unknown-linux-gnu
release: 1.71.1
LLVM version: 16.0.5
```

**Comment by bobhy** [comment]
See Cargo team analysis at: https://github.com/rust-lang/cargo/issues/12561#issuecomment-1693995710
